### ISSUE TITLE
Refactor data format conversion utilities into separate packages?

### ISSUE THREAD

**Issue by csbrown-noaa** [ultralytics/ultralytics#22159]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

I use the YOLO <-> COCO conversion utilities in the ultralytics library for some data ETL tasks.  While these tasks are related to AI model development, they are pretty lightweight, and happen entirely disjointed from any model training or prediction.  As such, it's a little unfortunate that I need to install a 1GB torch distribution just to munge some bounding box annotations.  Not sure if there's any other parts of this library that might benefit from this, but it would be cool if the ultralytics monolith was more modular, and I could just install a narrower set of utilities related to my use case.

I could be convinced to submit a PR, but I think the hard work here would be actually standing up a new repo, getting a new PyPi entry, deciding what bits and bobs could exist where, etc.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @csbrown-noaa, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will review and follow up soon.

Great suggestion on making data format converters more modular. To help us scope this feature request, could you share a bit more detail?
- Which converters do you rely on most (e.g., YOLO↔COCO, segmentation/pose variants)?
- Would a lightweight, dependency-minimized package (no torch) meet your needs, and what’s your acceptable dependency footprint?
- Do you prefer a Python API, CLI tools, or both?
- Any must-have features (streaming I/O, validation, schema checks, large dataset handling)?
- Are there specific constraints regarding Python versions or environments you need supported?

If you’re open to it, outlining a brief proposal (intended modules, minimal deps, API/CLI examples) would be very helpful for discussion. You can also start a broader design conversation with the community on [Discourse](https://community.ultralytics.com/) or in real time on [Discord](https://discord.com/invite/ultralytics) 🎧.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples that may be relevant.

If this is a custom training ❓ question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by csbrown-noaa** [comment]
- Which converters do you rely on most (e.g., YOLO↔COCO, segmentation/pose variants)?
    - I currently am only actively using the bounding boxes, but I also have a future need for segementation and arbitrary keypoints.
- Would a lightweight, dependency-minimized package (no torch) meet your needs, and what’s your acceptable dependency footprint?
    - "needs" is a strong word.  I think the dependencies for `ultralytics` amount to a few GB though, so any ability to reduce this depending on the specific workflow I am deploying would be nice.
- Do you prefer a Python API, CLI tools, or both?
    - None of the above.  All of the `ultralytics` API and CLI tools would remain unchanged.
- Any must-have features (streaming I/O, validation, schema checks, large dataset handling)?
    - No added features.  Just package-level restructuring.
- Are there specific constraints regarding Python versions or environments you need supported?
    - No.

**Comment by glenn-jocher** [comment]
Thanks for raising this—making the data converters usable without the full Torch stack is reasonable. As a short-term workaround, you can call the converters directly (torch-free) as shown in the docs; only `yolo_bbox2segment` pulls SAM/torch:

```python
from ultralytics.data.converter import convert_coco

convert_coco("path/to/coco/annotations/", use_segments=False, use_keypoints=False, cls91to80=True)
```

See the converter utilities reference for details: [Reference: ultralytics.data.converter](https://docs.ultralytics.com/reference/data/converter/). If you want to minimize dependencies today, consider the “install without dependencies” approach and add only what the converter functions require as noted here: [Custom installation methods](https://docs.ultralytics.com/quickstart/#custom-installation-methods).

For the proposal itself, we’re open to an incremental PR that keeps `ultralytics.data` import-light (no top-level Torch), adds a small “data” optional extra, and includes tests/docs scoped just to the converters; we’ll review without giving timelines.

**Comment by nikunjlad** [comment]
@csbrown-noaa you should try to check this pypi module : [datumaro](https://open-edge-platform.github.io/datumaro/stable/docs/get-started/introduction.html). It does what you need and is not relying on torch. If you are willing to use a different python package for your preprocessing stage data needs, and then use ultralytics, then datumaro can help you get that.

**Comment by csbrown-noaa** [comment]
@nikunjlad this is a good idea.  Do you if the ultralytics YOLO yaml "conforms" to whatever Datamuro expects?  Datamuro links to the darknet implementation of YOLO...
